### PR TITLE
chore: create message envelope for sending (AR-972)

### PR DESCRIPTION
**Pull request by vitorhugods**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] contains a reference JIRA issue number like `SQPIT-764`
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

We need to encrypt a message for each recipient.

### Solutions

`MessageEnvelopeCreator` takes a message and a list of recipients and is responsible for using whatever encryption mechanism needed in order to output a `MessageEnvelope`

### Testing

Covered `MessageEnvelopeCreator` with tests.

### Notes (Optional)

External messages are not yet supported (when content is too big and we only encrypt once).

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreator.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+) : MessageEnvelopeCreator {
+
+    override suspend fun createOutgoingEnvelope(
+        recipients: List<Recipient>,
+        message: Message
+    ): Either<CoreFailure, MessageEnvelope> = suspending {
+        val senderClientId = message.senderClientId
+        val content = protoContentMapper.encodeToProtobuf(ProtoContent(message.id, message.content))
+
+        recipients.foldToEitherWhileRight(mutableListOf<RecipientEntry>()) { recipient, recipientAccumulator ->
```
Comment (MohamadJaara) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreator.kt:41]: should user self clients be added to the recipients' list?

Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreator.kt:41]: First: Go to sleep! It's 1:38 in the morning 💤 

Second: Yes. It's not the `MessageEnvelopeCreator`'s responsibility to judge that. But there will be an entity that will get all recipients (including other self user clients) and call the envelope creator with that.

**Review by ohassine — APPROVED** [review]
👌

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapper.kt [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
 
-// IDE is currently complaining that it's not implemented on JVM. The compiler builds it correctly. The IDE is wrong.
-expect class ProtoContentMapper {
+interface ProtoContentMapper {
     fun encodeToProtobuf(protoContent: ProtoContent): PlainMessageBlob
     fun decodeFromProtobuf(encodedContent: PlainMessageBlob): ProtoContent
 }
+
+// IDE is currently complaining that it's not implemented on JVM. The compiler builds it correctly. The IDE is wrong.
+expect class PlatformProtoContentMapper: ProtoContentMapper
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapper.kt:9]: Why the `Platform`prefix? Wouldn't it be more consistent with `ProtoContentMapperImpl`?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageEnvelope.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package com.wire.kalium.logic.data.message
+
+import com.wire.kalium.logic.data.conversation.ClientId
+import com.wire.kalium.logic.data.user.UserId
+
+data class MessageEnvelope(
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageEnvelope.kt:6]: Note that this `MessageEnvelope` only makes sense for proteus.

Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageEnvelope.kt:6]: Yes. We can replace with an interface / sealed class later for different protocols.

Same should apply for a lot of the work being done now.

**Pull request merged**